src/detect-tls-cert-validity.c

```diff
@@ -0,0 +1,1331 @@
+            goto error;
+        }
+    }
+    return dd;
+
+error:
+    if (dd)
+        SCFree(dd);
+    return NULL;
+}
+
+/**
+ * \brief Function to add the parsed tls_notbefore into the current signature.
+ *
+ * \param de_ctx Pointer to the Detection Engine Context.
+ * \param s      Pointer to the Current Signature.
+ * \param rawstr Pointer to the user provided flags options.
+ *
+ * \retval 0 on Success.
+ * \retval -1 on Failure.
+ */
+static int DetectTlsNotBeforeSetup (DetectEngineCtx *de_ctx, Signature *s,
+                                    char *rawstr)
+{
+    uint8_t type = DETECT_TLS_TYPE_NOTBEFORE;
+    int r = DetectTlsValiditySetup(de_ctx, s, rawstr, type);
+
+    SCReturnInt(r);
+}
+
+/**
+ * \brief Function to add the parsed tls_notafter into the current signature.
+ *
+ * \param de_ctx Pointer to the Detection Engine Context.
+ * \param s      Pointer to the Current Signature.
+ * \param rawstr Pointer to the user provided flags options.
+ *
+ * \retval 0 on Success.
+ * \retval -1 on Failure.
+ */
+static int DetectTlsNotAfterSetup (DetectEngineCtx *de_ctx, Signature *s,
+                                    char *rawstr)
+{
+    uint8_t type = DETECT_TLS_TYPE_NOTAFTER;
+    int r = DetectTlsValiditySetup(de_ctx, s, rawstr, type);
+
+    SCReturnInt(r);
+}
+
+/**
+ * \brief Function to add the parsed tls validity field into the current signature.
+ *
+ * \param de_ctx Pointer to the Detection Engine Context.
+ * \param s      Pointer to the Current Signature.
+ * \param rawstr Pointer to the user provided flags options.
+ * \param type   Defines if this is notBefore or notAfter.
+ *
+ * \retval 0 on Success.
+ * \retval -1 on Failure.
+ */
+static int DetectTlsValiditySetup (DetectEngineCtx *de_ctx, Signature *s,
+                                   char *rawstr, uint8_t type)
+{
+    DetectTlsValidityData *dd = NULL;
+    SigMatch *sm = NULL;
+
+    SCLogDebug("\'%s\'", rawstr);
+
+    dd = DetectTlsValidityParse(rawstr);
+    if (dd == NULL) {
+        SCLogError(SC_ERR_INVALID_ARGUMENT,"Parsing \'%s\' failed", rawstr);
+        goto error;
+    }
+
+    /* okay so far so good, lets get this into a SigMatch
+     * and put it in the Signature. */
+    sm = SigMatchAlloc();
+    if (sm == NULL)
+        goto error;
+
+    if (s->alproto != ALPROTO_UNKNOWN && s->alproto != ALPROTO_TLS) {
+        SCLogError(SC_ERR_CONFLICTING_RULE_KEYWORDS,
+                   "rule contains conflicting keywords.");
+        goto error;
+    }
+
+    if (type == DETECT_TLS_TYPE_NOTBEFORE) {
+        dd->type = DETECT_TLS_TYPE_NOTBEFORE;
+        sm->type = DETECT_AL_TLS_NOTBEFORE;
+    }
+    else if (type == DETECT_TLS_TYPE_NOTAFTER) {
+        dd->type = DETECT_TLS_TYPE_NOTAFTER;
+        sm->type = DETECT_AL_TLS_NOTAFTER;
+    }
+    else {
+        goto error;
+    }
+
+    sm->ctx = (void *)dd;
+
+    s->flags |= SIG_FLAG_APPLAYER;
+    s->alproto = ALPROTO_TLS;
+
+    SigMatchAppendSMToList(s, sm, DETECT_SM_LIST_AMATCH);
```

Comment:
I'm trying to phase out AMATCH. Now that TLS uses TX (even if just one TX) I'd like to see it use the same logic as the DNS response (DE_STATE_FLAG_DNSRESPONSE_INSPECT). Pretty much a separate inspect engine for TLS responses. See DetectEngineInspectGenericList in detect-engine-dns.c